modules/thermal_hydraulics/src/materials/ADWallHeatTransferCoefficientGnielinskiMaterial.C

```diff
@@ -0,0 +1,81 @@
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADWallHeatTransferCoefficientGnielinskiMaterial.h"
+#include "SinglePhaseFluidProperties.h"
+#include "FlowModelSinglePhase.h"
+#include "Numerics.h"
+
+registerMooseObject("ThermalHydraulicsApp", ADWallHeatTransferCoefficientGnielinskiMaterial);
+
+InputParameters
+ADWallHeatTransferCoefficientGnielinskiMaterial::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addParam<MaterialPropertyName>("Hw",
+                                        FlowModelSinglePhase::HEAT_TRANSFER_COEFFICIENT_WALL,
+                                        "Heat transfer coefficient material property");
+  params.addParam<MaterialPropertyName>(
+      "rho", FlowModelSinglePhase::DENSITY, "Density of the fluid");
+  params.addParam<MaterialPropertyName>("vel", FlowModelSinglePhase::VELOCITY, "Fluid velocity");
+  params.addParam<MaterialPropertyName>(
+      "D_h", FlowModelSinglePhase::HYDRAULIC_DIAMETER, "Hydraulic diameter");
+  params.addParam<MaterialPropertyName>(
+      "cp", FlowModelSinglePhase::SPECIFIC_HEAT_CONSTANT_PRESSURE, "Specific heat of the fluid");
+  params.addParam<MaterialPropertyName>(
+      "mu", FlowModelSinglePhase::DYNAMIC_VISCOSITY, "Dynamic viscosity of the fluid");
+  params.addParam<MaterialPropertyName>(
+      "k", FlowModelSinglePhase::THERMAL_CONDUCTIVITY, "Heat conductivity of the fluid");
+  params.addParam<MaterialPropertyName>(
+      "T", FlowModelSinglePhase::TEMPERATURE, "Fluid temperature");
+  params.addParam<MaterialPropertyName>("T_wall", FlowModel::TEMPERATURE_WALL, "Wall temperature");
+  params.addClassDescription(
+      "Computes wall heat transfer coefficient for water using the Gnielinski correlation");
+  return params;
+}
+
+ADWallHeatTransferCoefficientGnielinskiMaterial::ADWallHeatTransferCoefficientGnielinskiMaterial(
+    const InputParameters & parameters)
+  : Material(parameters),
+    _Hw(declareADProperty<Real>("Hw")),
+    _rho(getADMaterialProperty<Real>("rho")),
+    _vel(getADMaterialProperty<Real>("vel")),
+    _D_h(getADMaterialProperty<Real>("D_h")),
+    _k(getADMaterialProperty<Real>("k")),
+    _mu(getADMaterialProperty<Real>("mu")),
+    _cp(getADMaterialProperty<Real>("cp")),
+    _T(getADMaterialProperty<Real>("T")),
+    _T_wall(getADMaterialProperty<Real>("T_wall"))
+{
+}
+
+void
+ADWallHeatTransferCoefficientGnielinskiMaterial::computeQpProperties()
+{
+  ADReal Pr = THM::Prandtl(_cp[_qp], _mu[_qp], _k[_qp]);
+  ADReal Re = std::max(1.0, THM::Reynolds(1., _rho[_qp], _vel[_qp], _D_h[_qp], _mu[_qp]));
+
+  if (Re < 1000)
+  {
+    mooseError("Gnielinski's correlation will produce negative heat transfer coefficients if the "
+               "Reynolds number is below 1000, please verify your assumptions");
```

Comment:
Make this a warning. Sometimes, we start a simulation with low velocities to help with convergence, but may still want to use that correlation.